language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -184,228 +126,471 @@ pub struct RootAccessPath {
 
 /// A GlobalRef holds the reference to the data and a shared reference to the root so
 /// status flags and reference count can be properly managed
-#[derive(Debug, Clone)]
+#[derive(PartialEq, Eq, Debug, Clone)]
 pub struct GlobalRef {
     root: Rc<RefCell<RootAccessPath>>,
     reference: MutVal,
 }
 
-impl Clone for MutVal {
-    fn clone(&self) -> Self {
-        MutVal(Rc::new(RefCell::new(self.peek().clone())))
+// All implementation here is private to this module and the effective logic in
+// working with values in the VM.
+impl ValueImpl {
+    fn into_value(self) -> VMResult<Value> {
+        match self {
+            ValueImpl::Invalid => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
+            ValueImpl::LocalRef(reference) => reference.into_value(),
+            _ => Ok(Value(self)),
+        }
     }
-}
 
-impl Clone for Local {
-    fn clone(&self) -> Self {
+    fn copy_value(&self) -> VMResult<Value> {
         match self {
-            Local::Ref(v) => Local::Ref(v.shallow_clone()),
-            Local::GlobalRef(v) => Local::GlobalRef(v.shallow_clone()),
-            Local::Value(v) => Local::Value(v.clone()),
-            Local::Invalid => Local::Invalid,
+            ValueImpl::Invalid => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
+            ValueImpl::LocalRef(reference) => reference.copy_value(),
+            _ => Ok(Value(self.clone())),
         }
     }
-}
 
-impl MutVal {
-    pub fn try_own(mv: Self) -> VMResult<Value> {
-        match Rc::try_unwrap(mv.0) {
-            Ok(cell) => Ok(cell.into_inner()),
-            Err(_) => Err(VMStatus::new(StatusCode::LOCAL_REFERENCE_ERROR)),
+    fn borrow_field(&self, field_offset: usize) -> VMResult<Value> {
+        match self {
+            ValueImpl::Struct(s) => s.get_field_reference(field_offset),
+            _ => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
         }
     }
 
-    pub fn peek(&self) -> Ref<Value> {
-        self.0.borrow()
+    fn size(&self) -> AbstractMemorySize<GasCarrier> {
+        match self {
+            ValueImpl::Invalid | ValueImpl::U64(_) | ValueImpl::Bool(_) => *CONST_SIZE,
+            ValueImpl::Address(_) => AbstractMemorySize::new(ADDRESS_LENGTH as u64),
+            // Possible debate topic: Should we charge based upon the size of the string.
```

Comment:
I think we should charge based on the size of the string.

Reply:
there is work to do for gas so I'll defer that to when we do that work with @tzakian.
Btw, this is a comment that was there already